usdigitalresponse/cpf-reporter
# Adds ability for developers to authenticate users in the local environment

## Context
With the implementation of [Passage](docs.passage.id) we are introducing a dependence on a third-party service to manage user-authentication. This can pose a challenge for local development if every developer needs to sign-up to get API-keys etc.
This PR ensures that in the local environment developers have the ability to authenticate a user merely by providing an email address of a valid user.


## How does the flow of data work?
1. `LoginPage` front-end component renders the appropriate page depending on whether the environment's auth-provider is `local | custom-auth | passage`. Note: `custom-auth` will be deprecated once `passage` is implemented.
<img width="744" alt="image" src="https://github.com/usdigitalresponse/cpf-reporter/assets/6497366/dfb04b51-29f8-4f5e-8fa6-e359772d4e73">
<img width="520" alt="image" src="https://github.com/usdigitalresponse/cpf-reporter/assets/6497366/7d15d1dc-be02-45fe-9af7-9ebeaf9bc1c7">


2. `web/src/auth.ts` instantiates the correct auth implementation client based on the same `AUTH_PROVIDER` environment variable.

3. The Auth client makes requests to `localhost:8911/localAuth` which is a new endpoint served by `api/src/functions/localAuth/localAuth.ts` function.

4. The `localAuth.ts` handler verifies the request and looks for the user-email submitted and checks whether a user with that email exists in the database. If the user is not found it sends a 4xx code or sends the user-object back in the 200 response.

5. The front-end stores the email the `local_auth_token` key of `localStorage`. This will then be passed in as part of the header in subsequent `localhost:8911/graphql` requests as `Authorization: Bearer ${localStorage.local_auth_token}`

6. The `createGraphQLHandler` then passes this information to the `getCurrentUser` function in  `api/src/lib/auth` which does a similar check as Step 4 and passes back the user that is tied to the token passed in the `Authorization` header.

## Testing
1. Ensure the database is seeded with the latest users added in this PR.
2. Ensure that the `AUTH_PROVIDER` environment variable is set to `local` and the app has restarted since changing this variable.
3. Navigate to `localhost:8910/login`
4. Verify that `foo@example.com` redirects back to the login page.
5. Verify that `grants-admin@usdigitalresponse.org` successfully redirects to `localhost:8910/`
6. Verify that hitting `Logout` on the top right redirects back to `/login`
7. Verify that going to root `/` redirects back to `/login` since no user is logged in
